lib/reek/configuration/app_configuration.rb

```diff
@@ -5,89 +5,115 @@ module Reek
+      # directory_directives = { Pathname("spec/samples/three_clean_files/") =>
+      #                          {Reek::Smells::UtilityFunction=>{"enabled"=>false}}}
+      #
+      #
+      # @param options [OpenStruct]
+      #   e.g.: #<OpenStruct report_format=:text,
+      #                      location_format=:numbers,
+      #                      colored=true,
+      #                      smells_to_detect=[],
+      #                      config_file=#<Pathname:config/defaults.reek>,
+      #                      argv=["lib/reek/spec"]>
+      def initialize(options = nil)
+        self.directory_directives = {}
+        self.default_directive    = {}
+        self.exclude_paths        = []
 
-        def initialize_with(options)
-          @has_been_initialized = true
-          configuration_file_path = ConfigurationFileFinder.find(options: options)
-          return unless configuration_file_path
-          load_from_file configuration_file_path
-        end
+        load options
+      end
 
-        def configure_smell_repository(smell_repository)
-          # Let users call this method directly without having initialized AppConfiguration before
-          # and if they do, initialize it without application context
-          initialize_with(nil) unless @has_been_initialized
-          for_smell_types.each do |klass_name, config|
-            klass = load_smell_type(klass_name)
-            smell_repository.configure(klass, config) if klass
-          end
-        end
+      # @param source_via [String] - the source of the code inspected
+      # @return [Hash] the directory directive for the source or, if there is
+      # none, the default directive
+      def directive_for(source_via)
+        directory_directive_for_source(source_via) || default_directive
+      end
 
-        def load_from_file(path)
-          if path.size.zero?
-            report_problem('Empty file', path)
-            return
-          end
+      private
 
-          begin
-            @configuration = YAML.load_file(path) || {}
-          rescue => error
-            raise_error(error.to_s, path)
-          end
+      attr_writer :exclude_paths, :default_directive, :directory_directives
 
-          raise_error('Not a hash', path) unless @configuration.is_a? Hash
-        end
+      # @param source_via [String] - the source of the code inspected
+      # Might be a string, STDIN or Filename / Pathname. We're only interested in the source
+      # when it's coming from file since a directory_directive doesn't make sense
+      # for anything else.
+      # @return [Hash | nil] the configuration for the source or nil
+      def directory_directive_for_source(source_via)
+        return unless source_via
+        base_path = Pathname.new(source_via).split[0]
+        hit = directory_directives.keys.detect { |pathname| pathname.to_s =~ /#{base_path}/ }
+        directory_directives[hit] unless hit.nil?
```

Comment:
How does this exactly work (i.e., why the regexp match)? Can you show an example of a match?

(The regexp match of two `Pathnames` cast to `Strings`, especially without the `base_path` going through `Regexp.escape`, worries me.)